state/remoteapplication.go

```diff
@@ -378,34 +435,51 @@ func (st *State) AddRemoteApplication(name, url string, endpoints []charm.Relati
 
 // RemoteApplication returns a remote application state by name.
 func (st *State) RemoteApplication(name string) (_ *RemoteApplication, err error) {
-	applications, closer := st.getCollection(remoteApplicationsC)
-	defer closer()
-
 	if !names.IsValidApplication(name) {
 		return nil, errors.NotValidf("remote application name %q", name)
 	}
-	sdoc := &remoteApplicationDoc{}
-	err = applications.FindId(name).One(sdoc)
+
+	applications, closer := st.getCollection(remoteApplicationsC)
+	defer closer()
+
+	appDoc := &remoteApplicationDoc{}
+	err = applications.FindId(name).One(appDoc)
 	if err == mgo.ErrNotFound {
 		return nil, errors.NotFoundf("remote application %q", name)
 	}
 	if err != nil {
 		return nil, errors.Annotatef(err, "cannot get remote application %q", name)
 	}
-	return newRemoteApplication(st, sdoc), nil
+	return newRemoteApplication(st, appDoc), nil
+}
+
+// RemoteApplicationByToken returns a remote application state by token.
+func (st *State) RemoteApplicationByToken(token string) (_ *RemoteApplication, err error) {
+	apps, closer := st.getCollection(remoteApplicationsC)
+	defer closer()
+
+	appDoc := &remoteApplicationDoc{}
+	err = apps.Find(bson.D{{"token", token}}).One(appDoc)
+	if err == mgo.ErrNotFound {
+		return nil, errors.NotFoundf("remote application with token %q", token)
+	}
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot get remote application with token %q", token)
+	}
+	return newRemoteApplication(st, appDoc), nil
 }
 
 // AllRemoteApplications returns all the remote applications used by the model.
 func (st *State) AllRemoteApplications() (applications []*RemoteApplication, err error) {
 	applicationsCollection, closer := st.getCollection(remoteApplicationsC)
 	defer closer()
 
-	appdocs := []remoteApplicationDoc{}
-	err = applicationsCollection.Find(bson.D{}).All(&appdocs)
+	appDocs := []remoteApplicationDoc{}
+	err = applicationsCollection.Find(bson.D{}).All(&appDocs)
 	if err != nil {
 		return nil, errors.Errorf("cannot get all remote applications")
 	}
-	for _, v := range appdocs {
+	for _, v := range appDocs {
```

Comment:
please make applications not be a named return and construct it as an array with len(appDocs) instead.